### ISSUE TITLE
FR: Account linking for 3rd party accounts, eg via OIDC

### ISSUE THREAD

**Issue by samip5** [Ylianst/MeshCentral#6621]
**Is your feature request related to a problem? Please describe.**
I'm really frustrated when I cannot link my external account to a already existing admin account

**Describe the solution you'd like**
I would like the ability to link them together so that I can login to the admin account with SSO

**Additional context**
This behavior is very common among apps that support SSO

**Comment by si458** [comment]
unfortunately this might not possible or very difficult to achieve
the user identifiers are unique `per domain` AND `per external auth`

for example local accounts use the format `user/domain/(user or email)`

where as external providers like OIDC use the format `user/domain/~oidc:(PROVIDED BY OIDC PROVIDER)`
or saml `user/domain/~saml:(PROVIDED BY SAML PROVIDER)`

so we have no control over the UUID that gets given by the external provider

the only provider that COULD be possible to merge together would be LDAP, because we dont use the `~saml` or `~oidc`
the is already a request https://github.com/Ylianst/MeshCentral/issues/6442

**Comment by samip5** [comment]
Some IdP's have multiple ways to influence the UUID provided, eg it can be a hash, it can be an upn etc so I think it could be possible.

**Comment by si458** [comment]
@samip5 yes sorry when i say UUID i mean the `.sub` that gets returned from OIDC

but also it is indeed possible to change the UUID!
for example my Authentik server generates the UUID as a UUID,
BUT its possible to change it to provide the UUID as an email address or name again!
i had to do this other day to fix an issue for a gent using saml https://github.com/Ylianst/MeshCentral/commit/a1854fa07428348632b696eaf289debfc454ea67 https://github.com/Ylianst/MeshCentral/issues/6596

what we could do is add an option into the OIDC/SAML section of config.json to say `use local accounts`

this way you create local users as `user/domain/bob@fred.com` or `user/domain/bob`

then when you login with your OIDC/SAML provider and its returns to meshcentral,
we can reformat the user id from `user/domain/~oidc:bob@fred.com` or `user/domain/~oidc:bob`
to `user/domain/bob@fred.com` or `user/domain/bob`

also one other thing to remember if OIDC/SAML have group sync, so u would create your user groups etc in OIDC/SAML
then get meshcentral to download all the groups, then ur admin account from the OIDC, logs into meshcentral and then gives the groups permissions etc to devices in, if that makes sense?

**Comment by samip5** [comment]
Yes, also using Authentik.

It would make sense to link accounts together if their email address matches or the ability to log-in with said account and then click a button to link which would do the oidc link.

**Comment by Fuseteam** [comment]
also using authentik, if we can match users using email that would be nice indeed. seems with authentik we can change the sub to email aswell if that helps

**Comment by samip5** [comment]
Stalebot go away.

**Comment by Fuseteam** [comment]
stalebot, no staling

**Comment by Fuseteam** [comment]
This is very much still an issue, as a the new users need to granted access to the same group as their local user

**Comment by Fuseteam** [comment]
down stalebot, down!

**Comment by Fuseteam** [comment]
no stalebot, bad bot

**Comment by si458** [comment]
Ok so will try have a look at this again next week!

Please can anyone explain this to me again in simple terms?

Do u basically want a local user `user//myusername` to be linked with external providers rather than the `user//~oidc:randomstringhereoremailaddress` so you can pre-populate user accounts ready to go etc?

**Comment by Fuseteam** [comment]
so basically we would like to have the two be linked so existing users don't get recreated when they log in using OIDC
i think if either username or email address matches (depending on what the provider sends as identifier) it should be treated as the same user

**Comment by si458** [comment]
@Fuseteam ok... so....

Would you want the local identifier as the primary one and then the oidc one for example as an extra one?

Also How would you link the user's together?

Would the user login with a local account `user//simon` via the user/pass form then go say into the my user page and then click say LINK and it diverts u to the oidc login stage and then returns u to the my user page again to show its been linked and what its other identifier would be?

Sorry just trying plan out HOW u would want them linking etc?

What about how would admins link the users?

**Comment by Fuseteam** [comment]
> Would you want the local identifier as the primary one and then the oidc one for example as an extra one?

that sounds reasonable

> Would the user login with a local account user//simon via the user/pass form then go say into the my user page and then click say LINK and it diverts u to the oidc login stage and then returns u to the my user page again to show its been linked

that....actually sounds fine 🤔 the other identifier could based on what the new users identifier is currently

> Sorry just trying plan out HOW u would want them linking etc?
>
> What about how would admins link the users?

Honestly i naively thought of it this way: 
- user logs in with oidc
- meshcentral notices the username already exists
- user gets logged in
- account is now linked to oidc

admins would control whether the existing users gets autolinked or not— if not the login is simply rejected when the username already exists

**Comment by si458** [comment]
@Fuseteam this is whats confusing me tho, how are u going to link the accounts automatically

`user//~oidc:randomstringhere` -> `user//simon`  ?
They are completely different

Or are u saying your oidc and local accounts are email addresses?

`user//~oidc:simon@mydomain.com` -> `user//simon@mydomain.com` 

If its email based, we could make sure the `usernameisemail: true` is set and if so then link etc based on email address

**Comment by samip5** [comment]
It can be UUID, or really anything and they can entirely be different. It should still be possible to link them together by telling mesh about the other identifier.

Using email addresses for anything like this is a bad idea IMO, as that can change and that would encounter issues.

**Comment by Fuseteam** [comment]
> [@Fuseteam](https://github.com/Fuseteam) this is whats confusing me tho, how are u going to link the accounts automatically
> 
> `user//~oidc:randomstringhere` -> `user//simon` ? They are completely different
> 

It is not a random string? it could be a UUID but it is unreasonable imo for a user to expect meshcentral to recognize that

in my particular case i use Authentik, which i can configure to send email adress or username so it can easily be

`user///~oidc:fuseteam` or `user///~oidc:fuseteam@mydomain.com`

> Or are u saying your oidc and local accounts are email addresses?
> 
> `user//~oidc:simon@mydomain.com` -> `user//simon@mydomain.com`
> 
> If its email based, we could make sure the `usernameisemail: true` is set and if so then link etc based on email address

in my case it is set to email address

but again as i said "i _naively_ though...."

**Comment by Fuseteam** [comment]
down boy, down

**Comment by Fuseteam** [comment]
stand down soldier

**Comment by samip5** [comment]
Damn it
